main.nf

```diff
@@ -167,46 +486,736 @@ process fastqc {
+
+    input:
+    set id, file(sr) from trimmed_reads_spades  
+
+    output:
+    set id, val("SPAdes"), file("${id}_graph.gfa") into assembly_graph_spades
+    set val("SPAdes"), val("$id"), file("${id}_scaffolds.fasta") into assembly_spades_to_quast
+    set val("SPAdes"), val("$id"), file("${id}_scaffolds.fasta"), file(sr) into assembly_spades_to_metabat
+    file("${id}_contigs.fasta")
+    file("${id}_log.txt")
+
+    when:
+    !params.singleEnd && !params.skip_spades
+     
+    script:
+    def maxmem = "${task.memory.toString().replaceAll(/[\sGB]/,'')}"
+    """
+    metaspades.py \
+        --threads "${task.cpus}" \
+        --memory "$maxmem" \
+        --pe1-1 ${sr[0]} \
+        --pe1-2 ${sr[1]} \
+        -o spades
+    cp spades/assembly_graph_with_scaffolds.gfa ${id}_graph.gfa
+    cp spades/scaffolds.fasta ${id}_scaffolds.fasta
+    cp spades/contigs.fasta ${id}_contigs.fasta
+    cp spades/spades.log ${id}_log.txt
+    """
+}
+
+
+process quast {
+    tag "$assembler-$sample"
+    publishDir "${params.outdir}/Assembly/$assembler", mode: 'copy'
+
+    input:
+    set val(assembler), val(sample), file(assembly) from assembly_spades_to_quast.mix(assembly_megahit_to_quast).mix(assembly_spadeshybrid_to_quast)
+
+    output:
+    file("${sample}_QC/*") into quast_results
+
+    when:
+    !params.skip_quast
+
+    script:
+    """
+    metaquast.py --threads "${task.cpus}" --rna-finding --max-ref-number 0 -l "${assembler}-${sample}" "${assembly}" -o "${sample}_QC"
+    """
+}
+
+
+
+/*
+ * STEP 3 - Binning
+ */
+process metabat {
+    tag "$assembler-$sample"
+    publishDir "${params.outdir}/", mode: 'copy',
+        saveAs: {filename -> (filename.indexOf(".bam") == -1 && filename.indexOf(".fastq.gz") == -1) ? "GenomeBinning/$filename" : null}
+
+    input:
+    set val(assembler), val(sample), file(assembly), file(reads) from assembly_spades_to_metabat.mix(assembly_megahit_to_metabat).mix(assembly_spadeshybrid_to_metabat)
+    val(min_size) from params.min_contig_size
+
+    output:
+    set val(assembler), val(sample), file("MetaBAT2/*") into metabat_bins mode flatten
+    set val(assembler), val(sample), file("MetaBAT2/*") into metabat_bins_for_cat
+    set val(assembler), val(sample), file("MetaBAT2/*"), file(reads) into metabat_bins_quast_bins
+
+    when:
+    !params.skip_binning
+
+    script:
+    def name = "${assembler}-${sample}"
+    if ( !params.singleEnd ) {
+        """
+        bowtie2-build --threads "${task.cpus}" "${assembly}" ref
+        bowtie2 -p "${task.cpus}" -x ref -1 "${reads[0]}" -2 "${reads[1]}" | \
+            samtools view -@ "${task.cpus}" -bS | \
+            samtools sort -@ "${task.cpus}" -o "${name}.bam"
+        samtools index "${name}.bam"
+        jgi_summarize_bam_contig_depths --outputDepth depth.txt "${name}.bam"
+        metabat2 -t "${task.cpus}" -i "${assembly}" -a depth.txt -o "MetaBAT2/${name}" -m ${min_size}
```

Comment:
I think the binning performance would improve a lot if one could use multiple sample bam files as input to metabat.  Is it too much for the first release? Also, moving the mapping out as an independent step would improve stability and cause less time consuming reruns.

Reply:
Done!!!